File: app/models/venta_detalle.rb
class VentaDetalle < ActiveRecord::Base
  belongs_to :venta
  belongs_to :producto
  belongs_to :promocion
  belongs_to :cotizacion
  belongs_to :moneda
  belongs_to :lote
  


  scope :ids, lambda { |id| where(:id => id) }

  attr_accessor :subtotal, :subtotal_sin_descuento, :multiplicar_cotizacion

  def subtotal
    stotal = precio * cantidad
    stotal = cotizar_subtotal(stotal)
    stotal -= descuento
    stotal
  end

  def costo
    cost = producto.precio_promedio * cantidad
    cost = cotizar_subtotal(cost)
    cost
  end

  def subtotal_sin_descuento
    cotizar_subtotal(precio * cantidad)
  end

  def cotizar_subtotal(subtotal)
    cotizado = subtotal
    if not monto_cotizacion.nil?
      if multiplicar_cotizacion
        cotizado *= monto_cotizacion
      else
        cotizado /= monto_cotizacion
      end
    end

    if venta.moneda.redondeo
      cotizado.round(0)
    else
      cotizado.round(2)
    end
  end

  def multiplicar_cotizacion
    venta.moneda.id == cotizacion.moneda_base.id
  end
end

File: app/assets/javascripts/models/venta_detalle.js
// for more details see: http://emberjs.com/guides/models/defining-models/

Bodega.VentaDetalle = DS.Model.extend({
  venta: DS.belongsTo('venta', { async: true }),
  producto: DS.belongsTo('producto'),
  promocion: DS.belongsTo('promocion'),
  cantidad: DS.attr('number', {defaultValue: 1}),
  precio: DS.attr('number', {defaultValue: 0}),
  descuento: DS.attr('number', {defaultValue: 0}),
  imei: DS.attr('string'),
  dirty: DS.attr('boolean', {defaultValue: false}),
  cotizacion: DS.belongsTo('cotizacion', {async:true}),
  montoCotizacion: DS.attr('number'),
  moneda: DS.belongsTo('moneda', {async:true}),
  caliente: DS.attr('boolean', {defaultValue: false}),
  loteId: DS.attr('string'),
  codigoLote: DS.attr('string'),

  subtotal: function() {
    var total = (this.get('precio') * this.get('cantidad')) - this.get('descuento');
    return total < 0 ? 0 : total;
  }.property('cantidad', 'precio', 'descuento'),

  // transcient
  subtotalCotizado: DS.attr('number', {defaultValue: 0}),
  multiplicarCotizacion: DS.attr('boolean')
});
